File: src/Tests/Behavioral/PointerCastSliceRange/PointerCastSliceRange.cs
namespace go;

using fmt = fmt_package;
using @unsafe = unsafe_package;

partial class main_package {

internal static unsafe void Main() {
    ref var arr = ref heap<array<nint>>(out var Ꮡarr);
    arr = new nint[]{10, 20, 30, 40}.array();
    var s = new slice<nint>(new ReadOnlySpan<nint>((nint*)(uintptr)(new @unsafe.Pointer(Ꮡarr)), 4));
    nint sum = 0;
    foreach (var (i, _) in s) {
        sum += i;
    }
    nint total = 0;
    foreach (var (_, v) in s) {
        total += v;
    }
    foreach (var (i, _) in s) {
        var ps = Ꮡ(s, i);
        ps.Value += 1;
    }
    fmt.Println(sum, total, s[0]);
    var ip = Ꮡarr;
    var back = (~(ж<ж<array<nint>>>)(uintptr)(@unsafe.Pointer.FromRef(ref (Ꮡ(ip)).Value))).Value;
    _ = back;
}

} // end main_package

File: src/Tests/Behavioral/PointerCastSliceRange/PointerCastSliceRange.go
package main

import (
	"fmt"
	"unsafe"
)

// A (*[N]T)(ptr)[:n] pointer-cast slice is emitted as a golib slice<T> (not a bare Span<T>), so it
// ranges as (index, element) and supports element-address Ꮡ(s, i) — mirrors runtime's printDebugLog
// (range + &state[i]) and os_windows (range over an unsafe []byte). This is a COMPILE + target guard.
// Here the source `&arr` is already `*[4]int`, so `(*[4]int)(unsafe.Pointer(&arr))` is a no-op IDENTITY
// that the converter collapses to the array box and slices directly (`(~Ꮡarr)[..4]`) — a real
// slice<T> over the array's own storage, no transient unsafe.Pointer=nuint round-trip. (A NON-identity
// pointer-cast slice, whose source is an unsafe.Pointer VALUE, keeps the Span-based emission; the
// stdlib exercises that shape.)
func main() {
	arr := [4]int{10, 20, 30, 40}
	s := (*[4]int)(unsafe.Pointer(&arr))[:4]

	sum := 0
	for i := range s { // index-only range over a pointer-cast slice
		sum += i
	}

	total := 0
	for _, v := range s { // value range
		total += v
	}

	for i := range s { // element-address through &s[i]
		ps := &s[i]
		*ps += 1
	}

	fmt.Println(sum, total, s[0])

	// census F13: a DOUBLE deref of a double-pointer reinterpret - reflect MapOf's
	// `**(**mapType)(unsafe.Pointer(&imap))` - the outer deref must parenthesize the
	// inner `~`-form deref before appending .Value, or the postfix binds into the cast
	// chain and reads the inner unsafe.Pointer slot (CS0029). The target element type
	// DIFFERS from the source (`**[2]int64` over a `**[4]int`) so it stays a genuine
	// reinterpret — an identical type would collapse to the box. Compile-shape only.
	ip := &arr
	back := **(**[2]int64)(unsafe.Pointer(&ip))
	_ = back

	// census F14: a literal with an unsafe.Pointer result whose arms return different
	// C#-typed expressions defeats lambda return inference (CS8917, reflect deepEqual
	// ptrval); the emitted lambda states its return type explicitly.
	pick := func(u bool) unsafe.Pointer {
		if u {
			return unsafe.Pointer(uintptr(0))
		}
		return unsafe.Pointer(ip)
	}
	_ = pick(true)

	// A NAMED-over-pointer conversion from a raw address (syscall.Pointer shape,
	// internal/poll WSAMsg.Name CS0030 x6): hops uintptr then the named operator -
	// ((opaque)(zh<nint>)(uintptr)(...)). Compile-shape only.
	op := opaque(unsafe.Pointer(ip))
	_ = op
}

// opaque mirrors syscall.Pointer: a named type whose underlying is a pointer.
type opaque *[4]int

